### PR TITLE
RTL-377: add change streams for dynamic pricing

### PR DESCRIPTION
**Pull request by AngelicaGE**
Jira ticket: https://jira.mongodb.org/browse/RTL-377

Changes include
- Add change streams
- update pred price component
- add example .env file to include new vars needed for changestream

### CODE REVIEW

**Comment by AngelicaGE** [comment]
https://github.com/user-attachments/assets/c90ac86e-0262-45b3-9ce2-7495af8f5352

**Review by frankbaldi44 — APPROVED** [review]
LGTM

File: EXAMPEL.env.local
Comment (frankbaldi44) [inline EXAMPEL.env.local:1]: Typo in name of file. Please update.

File: app/api/sse/route.js
Comment (frankbaldi44) [inline app/api/sse/route.js:1]: @AngelicaGE  Can we add error handling? 


```
changeStream.on('error', (error) => {
--
console.error('Change stream error:', error);
writer.close(); // Close the writer on error
});
```

Comment (frankbaldi44) [inline app/api/sse/route.js:1]: Hi @AngelicaGE looks good, but let's make it a bit more robust:

- Handle Change Stream Closure: Let's listen to the change stream if it get's unexpectedly closed re-establish connection.

- Implement a Retry Mechanism: Use a simple backoff strategy to retry connecting to the change stream after a certain interval if it fails.

- Graceful Shutdown: Ensure that when the server is shutting down, you clean up resources properly.

**My suggestion to test (run the generator.py file to test the whole system)**

````
import { getPredPriceChangeStream } from '@/app/_lib/mongodb';
--
import { NextResponse } from 'next/server';
 
const HEARTBEAT_INTERVAL = 5000; // Keep alive interval in milliseconds
const RETRY_INTERVAL = 1000; // Initial retry interval in milliseconds
 
let changeStream; // Global variable to hold the change stream
let retryTimeout; // Variable to hold the retry timeout
 
async function setupChangeStream(writer) {
try {
changeStream = await getPredPriceChangeStream();
 
changeStream.on('change', (change) => {
// Notify the client about the change
sendUpdate(change, writer);
});
 
changeStream.on('error', (error) => {
console.error('Change stream error:', error);
// Attempt to re-establish the change stream
retryChangeStream(writer);
});
 
changeStream.on('close', () => {
console.log('Change stream closed, attempting to reconnect...');
retryChangeStream(writer);
});
} catch (error) {
console.error('Error setting up change stream:', error);
retryChangeStream(writer);
}
}
 
function retryChangeStream(writer) {
if (retryTimeout) return; // Prevent multiple retries
 
retryTimeout = setTimeout(async () => {
console.log('Retrying to set up change stream...');
await setupChangeStream(writer);
retryTimeout = null; // Reset the timeout
}, RETRY_INTERVAL);
}
 
function sendUpdate(data, writer) {
if (writer.locked) {
const event = `data: ${JSON.stringify(data)}\n\n`;
writer.write(new TextEncoder().encode(event)).catch((error) => {
console.error('Error writing update:', error);
});
} else {
console.warn('Writable stream is not locked, skipping update.');
}
}
 
export async function GET(req) {
if (req.headers.get('accept') === 'text/event-stream') {
const { readable, writable } = new TransformStream();
const writer = writable.getWriter();
 
// Set SSE headers
const headers = new Headers({
'Content-Type': 'text/event-stream',
'Cache-Control': 'no-cache',
'Connection': 'keep-alive',
});
 
const intervalId = setInterval(() => {
if (writer.locked) {
writer.write(new TextEncoder().encode(': heartbeat\n\n')).catch((error) => {
console.error('Error writing heartbeat:', error);
});
}
}, HEARTBEAT_INTERVAL);
 
// Set up the change stream
await setupChangeStream(writer);
 
// Handle client disconnect
req.signal.addEventListener('abort', () => {
console.log("Client disconnected");
clearInterval(intervalId);
if (changeStream) {
changeStream.close(); // Close the change stream
}
writer.close().catch((error) => {
console.error('Error closing writer:', error);
});
});
 
return new NextResponse(readable, { headers });
} else {
console.log('404');
return new NextResponse('Not Found', { status: 404 });
}
}
 
// Graceful shutdown for the server
process.on('SIGINT', () => {
console.log('Shutting down gracefully...');
if (changeStream) {
changeStream.close(); // Close the change stream
}
process.exit(0);
});

**Review by frankbaldi44 — CHANGES_REQUESTED** [review]
@AngelicaGE 

1. Please review filename typo
2. I've suggested some changes in the SSE/route.js
3. Let's test by running generator.py

File: EXAMPEL.env.local [hunk 1]
```diff
@@ -0,0 +1,8 @@
+IST_SHARED_MONGODB=
+GCP_PROJECT_ID=
+GCP_STORAGE_BUCKET=
+VERTEX_AI_ENDPOINT=
+
+MONGODB_URI=
```
Comment (bereilhp) [inline EXAMPEL.env.local:6]: Why do we have two MongoDB connection URIs (`IST_SHARED_MONGODB` and `MONGODB_URI`)? Can we consolidate them into a single URI called `MONGODB_URI`?

**Review by bereilhp — APPROVED** [review]
LGTM!

**Pull request merged**